analysis/src/xml.rs

```diff
@@ -0,0 +1,715 @@
+use roxmltree::StringStorage;
+use std::borrow::Cow;
+
+/// A node with its `'input` lifetime set to `'static`.
+type Node<'a> = roxmltree::Node<'a, 'static>;
+/// String type used for XML attribute and text values.
+///
+/// A `&'static str` is not used directly because sometimes string allocation is
+/// needed, for example when replacing `&quot;` with normal quotes.
+type XmlStr = Cow<'static, str>;
+
+/// Retrieves the value of the `node`'s attribute named `name`.
+fn attribute(node: Node, name: &str) -> Option<XmlStr> {
+    node.attribute_node(name)
+        .map(|attr| match attr.value_storage().clone() {
+            StringStorage::Borrowed(s) => Cow::Borrowed(s),
+            StringStorage::Owned(s) => Cow::Owned((*s).into()),
+        })
+}
+
+/// Retrieves the text inside the next child element of `node` named `name`.
+fn child_text(node: Node, name: &str) -> Option<XmlStr> {
```

Comment:
Does `descendant_text()` probably do the same, and is this function only a simpler (borrowing) variant that assumes there's never extra (text/comment) elements within the given `name` descendant?

Reply:
There's a difference between children and descendants in the sense that "children" only describe the nodes **directly** below the given one, whereas "descendants" will give you every children _and_ their sub-children.

This function will find the first direct child of the given node, and will then return you its text content in borrowed form. With "text content" I am referring to the node's first text child (which is what roxmltree's `.text_storage()` will give you). This is why there is no concern for extra comment elements.

This is all a bit unclear, we should add more comments to this code.

Reply:
Sure, the point was more so that `child_text()` is only intended for simple nodes that contain some text and no descendant nodes with text nor comments (and if they do, that is omitted). In that sense calling `descendant_text()` on the `child` node found here returns the same value. But the simplicity and borrowing nature of this function, together with the ability to select a child element of `node` probably warrants its existence.

(Note: I did check that calling `descendant_text(child)` is always identical to `child.text_storage()` here)